xdsl/dialects/gpu.py

```diff
@@ -66,85 +68,43 @@ class AsyncTokenType(ParametrizedAttribute, TypeAttribute):
     name = "gpu.async.token"
 
 
-@irdl_attr_definition
-class _AllReduceOperationAttr(ParametrizedAttribute):
-    name = "all_reduce_op"
+class AllReduceOpAttr(
+    Data[Literal["add", "and", "max", "min", "mul", "or", "xor"]], OpaqueSyntaxAttribute
+):
+    name = "gpu.all_reduce_op"
 
     param: ParameterDef[StringAttr]
 
-    def print_parameters(self, printer: Printer) -> None:
-        printer.print(f"all_reduce_op {self.param.data}")
-
-
-@irdl_attr_definition
-class _DimensionAttr(ParametrizedAttribute):
-    name = "dim"
-
-    param: ParameterDef[StringAttr]
-
-    def print_parameters(self, printer: Printer) -> None:
-        printer.print(f"dim {self.param.data}")
-
-
-T = TypeVar("T", bound=_AllReduceOperationAttr | _DimensionAttr, covariant=True)
-
-
-@irdl_attr_definition
-class _GPUAttr(ParametrizedAttribute, Generic[T]):
-    name = "gpu"
-
-    value: ParameterDef[T]
+    def print_parameter(self, printer: Printer) -> None:
+        printer.print(f" {self.data}")
 
     @classmethod
-    def parse_parameters(cls, parser: AttrParser) -> list[Attribute]:
-        parser.parse_characters(
-            "<",
-            ": gpu attributes currently have the #gpu<name value> syntax.",
-        )
-        if parser.parse_optional_keyword("dim"):
-            attrtype = _DimensionAttr
-            vtok = parser.parse_optional_identifier()
-            if vtok not in ["x", "y", "z"]:
-                parser.raise_error(
-                    f"Unexpected dim {vtok}. A gpu dim can only be x, y, or z",
-                )
+    def parse_parameter(
+        cls, parser: AttrParser
+    ) -> Literal["add", "and", "max", "min", "mul", "or", "xor"]:
+        for kw in ("add", "and", "max", "min", "mul", "or", "xor"):
+            val = parser.parse_optional_keyword(kw)
```

Comment:
same comment as previous PR, can we peek ahead here instead of parsing k times in a loop?

Reply:
Moreover, if the implementation could be the same for the parsing of both attributes, it'd be even better.

Reply:
We can just "parse_bare_ident", and then compare it to the elements in the list.

Reply:
@compor: The is definitely the next step ;) I want to implement a litlle helper for enum attributes, and switch the GPU ones to this.